389 - UI component - Configure a sequence

Francois389 opened the pull request:
close #389 

Ajout d'un nouveau composant Vue pour la configuration d'une séquence.
Intégration dans le template Thymleaf.

Mise à jour de Storybook vers la version 9.

Ajout d'un nouveau workflow pour Chromatic.

Francois389 committed:
feat: add chromatic configuration

- enable launching test with the UI

Francois389 committed:
chore: update storybook to v9

Francois389 committed:
feat: add codegen addon

Francois389 committed:
feat: add tag addon

Francois389 committed:
feat: add tag to existing component

Francois389 committed:
feat: ConfigurationSequence component + stories

Francois389 committed:
feat: add notice to ExecutionContext

+ add default ExecutionContext

Francois389 committed:
feat: add ConfrontingViews Phase setting

Francois389 committed:
feat: add Link component

Francois389 committed:
feat: ConfigurationSequence

Francois389 committed:
feat: export SequenceConfiguration component

Francois389 committed:
chore: ui-components v1.0.3

Francois389 committed:
feat: add test page to integrate SequenceConfiguration component

Francois389 committed:
chore: ui-components v1.0.4

- update Emit value

Francois389 committed:
chore: rename sequence-config to ui-sequence-config

Francois389 committed:
feat: add ui-component to template

+ update loader

Francois389 committed:
feat: start sequence with ui-componenent

Francois389 committed:
fix: the modal height is correctly set

Francois389 committed:
feat: add chromatic ci

Francois389 committed:
fix(ci): execute command in the right directory

Francois389 committed:
fix(ci): use correct parameter

Francois389 committed:
feat: add props to toggle AI explanation setting

Francois389 committed:
feat: initialise props in Thymleaf

jtranier reviewed (changes requested):
Au niveau interface c'est top !

Un tout petit problème, l'écran de chargement n'est pas bon ; ça met un loader centré sur la page avec une popup qui a une hauteur d'une ligne. Du coup le loader est en dehors...
Question : est-ce que l'on ne pourrait pas faire une story qui affiche l'état de chargement de la popup ?

Un petit détail dans le nommage en camlCase. Je ne l'aurais pas signalé s'il n'y avait que ça, mais à faire quelques changements... La convention veut que l'on commence chaque mot par une majuscule et que l'on ne mette que la première lettre des acronymes en majuscule. Du coup c'est `formatXml` et pas `formatXML`. Et pour IA c'est `Ia` et pas `IA`...

Problème fonctionne important par contre : il y a des contraintes subtiles sur les contextes. Par exemple, la fourniture de l'explication ne peut être enlevée qu'en mode face à face. Je ne sais pas où sont décrit ces exigences. @FranckSilvestre est-ce que tu as la réponse ? 
Car il est important de rester aligner sur le fonctionnement actuel.

Francois389 commented:
> Un tout petit problème, l'écran de chargement n'est pas bon ; ça met un loader centré sur la page avec une popup qui a une hauteur d'une ligne. Du coup le loader est en dehors... Question : est-ce que l'on ne pourrait pas faire une story qui affiche l'état de chargement de la popup ?

Ce qui ce passe avec les modals est un peu complexe.
Tout d'abord vu que la modal est affiché depuis une iframe, on envoie un evenement que le parent écoute, fetch le html de la modal ( premier loader centré).
Ensuite quand le html est chargé il y a un loader pour l'application Vue.

Et enfin j'ai tenté de d'utiliser semantic 2.3, mais ça causé des problèmes d'interfaces partout, et j'ai donc fais machine arrière. Donc pour que le contenue de la modal s'affiche en entier je l'ai centré manuellement en haut. C'est pour ça que ce n'est pas aligné.

jtranier committed:
chore: Make the code more declarative for greater clarity

jtranier commented:
@Francois389 I've replaced 
```
$("#start-sequence-" + sequenceId).on("click", function () { ... } 
```

with 
```
<a th:onclick="|elaastic.startSequence(${commandModel.sequenceId})|">
```

which much more readable (it attaches directly the behavior on the link)

jtranier committed:
chore: Compliance with URL conventions in SequenceController.kt

jtranier commented:
@Francois389 By convention the URL for the action `act` of the `XxxController` is `/xxx/act`.
This is useful because it allows to find easily the corresponding code from the URL. 
Otherwise, we to do it with a search text...

jtranier committed:
chore: Start a sequence posting a JSON object

jtranier commented:
I notice that there is a javascript error when using the popup. It complains jQuery is not defined. I assume it has not been loaded where the semantic popup is used, and that semantic makes use of it to implement the popup.

Anyway, I'm working on it. 
The code for starting a sequence seems to me much more complex than required (even if it is very well coded).

jtranier commented:
When starting a sequence, this error occurs in the console : 
```
Unknown command: sequence-configuration-submitted
```

It's hidden by the fact there is immediately a page reload afterwards

jtranier commented:
In this last commit, I've added a copy of the start sequence action in the player controller that expects a POST request (which makes more sense because we are updating the sequence). Hmmm... it could be a PUT by the way... 
And, moreover, it expects the data to be transmitted as a JSON object.

I've quickly rewrite the request mechanism to send this POST request with a JSON body.
This is just a 1st step for moving toward the VueJS component talking directly to the server, instead of adding this HTML form in the middle (which adds a lot of complexity)

jtranier committed:
fix: Add missing jQuery lib on vuejs-app.html layout

jtranier committed:
chore: Remove default case modal manager message handler

Many messages sent on the "window" do not concern the modal manager ; it is unnecessary to indicate it cannot interpret the message.

jtranier commented:
@Francois389 We have quite a lot of messages sent to `window` now. 
I think it would be a good idea to structure the messages ID, to scope them, and avoir conflicts.

e.g. messages ID could be in the form of `<component>/<message-id>`

Doing this, it'll make sense that the "Modal manager" signals an erreor when it does not know a message like `modal-manager/something`.

I've also seen there are constants for messages in Modal Manager, but they are used only there... 
It would OK to use the constant everywhere to facilitate refactoring.
Otherwise it is also acceptable to use direct string (if they not too much simple).
But I do not see the point of having constants without using it everywhere.

jtranier committed:
chore: Remove debug console.log statements from modal and sequence files.

Cleaned up unnecessary debug logs to improve code readability and reduce clutter in the browser console. Retained an error log for better error tracking during modal fetching.

jtranier committed:
chore: Simplify sequence configuration logic and remove form intermediation

Removed unnecessary form elements and logic for handling sequence configuration in Thymeleaf. Adjusted the modal behavior to directly communicate with the server using AJAX, prepping for improved UI-server interaction without intermediate layers. Added TODOs for future enhancements regarding data format and CSRF token handling.

jtranier commented:
@Francois389 J'ai avancé ce que je pouvais sur cette PR aujourd'hui ; je te repasse le relais.

Le problème de non gestion des règles de config des séquences vient du fait que la fonction `manageConfigurationChange` du fichier `_config-sequence.html` a été perdu en route.
Il faut bien sûr rétablir ces règles. Je les mettrais dans le UI Component. Il est fort probable que l'on puisse coder cela plus proprement en VueJS (je n'ai pas du tout eu le temps de regarder).

Je viens sur le dernier commit de virer le formulaire qui n'est plus utilisé & j'ai également supprimer l'ancienne action de contrôleur pour lancer la séquence (qui est remplacée par la nouvelle qui traite du JSON).

J'ai laissé 2 TODOs.
Il serait bien de traiter au moins le 1er, à savoir, harmoniser les données qu'envoie le composant de config de séquence pour qu'elles puissent être transmises telles quelles au serveur.

Le 2ème pourra faire l'objet d'une autre fiche. Cela va consister à mettre en place le pattern qui va bien pour que les UI composants puissent communiquer directement avec le serveur quand celui-ci propose des actions qui peuvent répondre du JSON. C'est essentiellement la gestion du jeton CSRF qu'il faut mettre en place.

A noter que, pour aller vraiment au bout du process, il faudrait que : 
- le serveur fasse une réponse JSON pour indiquer le résultat de l'action (la séquence a bien démarrée, ou... la description de l'erreur qui s'est produite)
- le client devrait prendre en compte une erreur éventuelle.
On verra ce dernier point plus tard.

Francois389 committed:
feat: update logic for student give an explanation

If the question is open, then the student must give an explanation.

If the execution context selected is Distance or Blended, then the student must give an explanation.

Francois389 committed:
test: add test to `sequence/{sequenceId}/start.json` endpoint

Francois389 committed:
chore: ui-component v1.0.6

Francois389 committed:
feat: update thymleaf template to give the necessary props to ui component

Francois389 commented:
> Le problème de non gestion des règles de config des séquences vient du fait que la fonction `manageConfigurationChange` du fichier `_config-sequence.html` a été perdu en route. Il faut bien sûr rétablir ces règles. Je les mettrais dans le UI Component. Il est fort probable que l'on puisse coder cela plus proprement en VueJS (je n'ai pas du tout eu le temps de regarder).
> 

J'ai rajouté les règles avec de35d5c153bc30180c63697a5e8bb9faf757266a. 
J'ai regardé sur la préprod comment ça se comportait.

Francois389 committed:
fix: restore PrincipalUserResolver).elaastic to get the authentificated user

Francois389 committed:
feat: update event to emit a single object we can send to the server

Francois389 committed:
feat: clean code

Francois389 committed:
chore: update url

jtranier committed:
chore: improve type definitions and constant naming

Updated `ExecutionContext` and `EvaluationMethod` types for clarity and defined them with specific string literals. Renamed constants to follow naming conventions, enhancing code readability and consistency.

jtranier committed:
chore: Remove "props." within SequenceConfiguration.vue template

jtranier committed:
chore: Minor code simplification

Francois389 committed:
feat: add missing confidence property to stories

jtranier commented on server/src/main/kotlin/org/elaastic/player/PlayerController.kt:
```diff
@@ -635,6 +631,8 @@ class PlayerController(
         @PathVariable responseId: Long
     ): String {
+        authentication.principal as User
+
```
Ce n'est plus le bon moyen de récupérer l'utilisateur authentifié. 
En faisant une recherche dans le code source, on retrouve pas mal d'occurence. Erreur de rebase ?

Francois389 replied:
Corrigé avec 602562600cbd42942aae4c721f6a8345c248e8ab

jtranier reviewed (changes requested):
Je n'ai pas relevé d'autres point que le pb de récupération d'utilisateur

Francois389 committed:
fix: get authentificate user with PrincipalUserResolver
